README.md
# S3 Object Streams

A small Node.js package that can be helpful when performing operations on very
large S3 buckets, those containing millions of objects or more, or when
processing S3 Inventory listings for those buckets. Streaming the listed
contents keeps memory under control and using the Streams API allows for fairly
compact utility code.

For very large buckets, S3 Inventory is always a better choice.

  * [S3ListObjectStream](#s3listobjectstream)
  * [S3ConcurrentListObjectStream](#s3concurrentlistobjectstream)
  * [S3UsageStream](#s3usagestream)
  * [S3InventoryUsageStream](#s3inventoryusagestream)

## Installing

Obtain via NPM:

```
npm install s3-object-streams
```

## S3ListObjectStream

An object stream that pipes in configuration objects for listing the contents of
an S3 bucket, and pipes out S3 object definitions.

```js
var AWS = require('aws-sdk');
var s3ObjectStreams = require('s3-object-streams');

var s3ListObjectStream = new s3ObjectStreams.S3ListObjectStream();
var s3Client = new AWS.S3();

// Log all of the listed objects.
s3ListObjectStream.on('data', function (s3Object) {
  console.info(s3Object);
});
s3ListObjectStream.on('end', function () {
  console.info('Listing complete.')
});
s3ListObjectStream.on('error', function (error) {
  console.error(error);
});

// List the contents of a couple of different buckets.
s3ListObjectStream.write({
  s3Client: s3Client,
  bucket: 'exampleBucket1',
  // Optional, only list keys with the given prefix.
  prefix: 'examplePrefix/',
  // Optional, group keys using a delimiter.
  // delimiter: '/',
  // Optional, defaults to 1000. The number of objects per request.
  maxKeys: 1000
});
s3ListObjectStream.write({
  s3Client: s3Client,
  bucket: 'exampleBucket2'
});
s3ListObjectStream.end();
```

Objects emitted by the stream have the standard format, with the addition of a
`Bucket` property:

```js
{
  Bucket: 'exampleBucket1',
  Key: 'examplePrefix/file.txt',
  LastModified: Date.now(),
  ETag: 'tag string',
  Size: 200,
  StorageClass: 'STANDARD',
  Owner: {
    DisplayName: 'exampleowner',
    ID: 'owner ID'
  }
```

## S3ConcurrentListObjectStream

This works in the same way as the `S3ListObjectStream`, but under the hood it
splits up the bucket by common prefixes and then recursively lists objects under
each common prefix concurrently, up to the maximum specified concurrency.

```js
var AWS = require('aws-sdk');
var s3ObjectStreams = require('s3-object-streams');

var s3ConcurrentListObjectStream = new s3ObjectStreams.S3ConcurrentListObjectStream({
  // Optional, defaults to 15.
  maxConcurrency: 15
});
var s3Client = new AWS.S3();

// Log all of the listed objects.
s3ConcurrentListObjectStream.on('data', function (s3Object) {
  console.info(s3Object);
});
s3ConcurrentListObjectStream.on('end', function () {
  console.info('Listing complete.')
});
s3ConcurrentListObjectStream.on('error', function (error) {
  console.error(error);
});

// List the contents of a couple of different buckets.
s3ConcurrentListObjectStream.write({
  s3Client: s3Client,
  bucket: 'exampleBucket1',
  // Optional, only list keys with the given prefix.
  prefix: 'examplePrefix/',
  // Optional, group keys using a delimiter.
  // delimiter: '/',
  // Optional, defaults to 1000. The number of objects per request.
  maxKeys: 1000
});
s3ConcurrentListObjectStream.write({
  s3Client: s3Client,
  bucket: 'exampleBucket2'
});
s3ConcurrentListObjectStream.end();
```

Objects emitted by the stream have the standard format, with the addition of a
`Bucket` property:

```js
{
  Bucket: 'exampleBucket1',
  Key: 'examplePrefix/file.txt',
  LastModified: Date.now(),
  ETag: 'tag string',
  Size: 200,
  StorageClass: 'STANDARD',
  Owner: {
    DisplayName: 'exampleowner',
    ID: 'owner ID'
  }
```

## S3UsageStream

A stream for keeping a running total of count and size of listed S3 objects by
bucket and key prefix. Useful for applications with a UI that needs to track
progress.

```js
var AWS = require('aws-sdk');
var s3ObjectStreams = require('s3-object-streams');

var s3ListObjectStream = new s3ObjectStreams.S3ListObjectStream();
var s3UsageStream = new s3ObjectStreams.S3UsageStream({
  // Determine folders from keys with this delimiter.
  delimiter: '/',
  // Group one level deep into the folders.
  depth: 1,
  // Only send a running total once every 100 objects.
  outputFactor: 100
});
var s3Client = new AWS.S3();

s3ListObjectStream.pipe(s3UsageStream);

var runningTotals;

// Log all of the listed objects.
s3UsageStream.on('data', function (totals) {
  runningTotals = totals;
  console.info(runningTotals);
});
s3UsageStream.on('end', function () {
  console.info('Final total: ', runningTotals);
});
s3UsageStream.on('error', function (error) {
  console.error(error);
});

// Obtain the total usage for these two buckets.
s3ListObjectStream.write({
  s3Client: s3Client,
  bucket: 'exampleBucket1'
});
s3ListObjectStream.write({
  s3Client: s3Client,
  bucket: 'exampleBucket2'
});
s3ListObjectStream.end();
```

The running total objects emitted by the stream have the following format:

```js
{
  path: 'exampleBucket/folder1',
  storageClass: {
    STANDARD: {
      // The number of files of this storage class.
      count: 55,
      // Total size in bytes of files in this storage class.
      size: 1232983
    },
    STANDARD_IA: {
      count: 0,
      size: 0
    },
    REDUCED_REDUNDANCY: {
      count: 2,
      size: 5638
    },
    GLACIER: {
      count: 0,
      size: 0
    }
  }
}
```

## S3InventoryUsageStream

A stream for keeping a running total of count and size of S3 objects by bucket
and key prefix, accepting objects from an S3 Inventory CSV file rather than
from the `listObjects` API endpoint.

```js
// Core.
var fs = require('fs');
var zlib = require('zlib');

// NPM.
var csv = require('csv');
var _ = require('lodash');
var s3ObjectStreams = require('s3-object-streams');

// Assuming that we already have the manifest JSON and a gzipped CSV data file
// downloaded from S3:
var manifest = require('/path/to/manifest.json');
var readStream = fs.createReadStream('/path/to/data.csv.gz');

var s3InventoryUsageStream = new s3ObjectStreams.S3InventoryUsageStream({
  // Determine folders from keys with this delimiter.
  delimiter: '/',
  // Group one level deep into the folders.
  depth: 1,
  // Only send a running total once every 100 objects.
  outputFactor: 100
});

var runningTotals;

// Log all of the listed objects.
s3UsageStream.on('data', function (totals) {
  runningTotals = totals;
  console.info(JSON.stringify(runningTotals, null, '  '));
});

var complete = _.once(function (error) {
  if (error) {
    console.error(error);
  }
  else {
    console.info('Complete');
  }
});

var gunzip = zlib.createGunzip();
var csvParser = csv.parse({
  // The manifest file defines the columns for the CSV. Specifying them here
  // ensures that the parser returns objects with properties named for the
  // columns.
  columns: manifest.fileSchema.split(', ')
});

csvParser.on('error', complete);
gunzip.on('error', complete);
readStream.on('error', complete);
s3InventoryUsageStream.on('error', complete);
s3InventoryUsageStream.on('end', complete);

// Unzip the file on the fly, feed it to the csvParser, and then into the
// object stream.
readStream.pipe(gunzip).pipe(csvParser).pipe(transformer).pipe(s3InventoryUsageStream);
```

The running total objects emitted by the stream have the following format:

```js
{
  path: 'exampleBucket/folder1',
  storageClass: {
    STANDARD: {
      // The number of files of this storage class.
      count: 55,
      // Total size in bytes of files in this storage class.
      size: 1232983
    },
    STANDARD_IA: {
      count: 0,
      size: 0
    },
    REDUCED_REDUNDANCY: {
      count: 2,
      size: 5638
    },
    GLACIER: {
      count: 0,
      size: 0
    }
  }
}
```


CHANGELOG.md
# Changelog

## 0.8.0

  * Add delimiter option to S3ListObjectStream class.

## 0.7.0

  * Add S3InventoryUsage class.

## 0.6.2

  * Fix possible problem resulting from mistaken assumption about API behavior.

## 0.6.1

  * Add missing STANDARD_IA storage class.

## 0.6.0

  * A better algorithm for concurrent listing.

## 0.5.0

  * Split out usage counts by storage class.

## 0.4.2

  * Further fixing of the S3ConcurrentListObjectStream.

## 0.4.1

  * Fix issue with passing in `prefix/` rather than `prefix`.

## 0.4.0

  * Add retry on error to S3 API operations.

## 0.3.0

  * Add the S3ConcurrentListObjectStream class.

## 0.2.0

  * Add an outputFactor configuration parameter to S3UsageStream.

## 0.1.1

  * Fix bug in S3UsageStream.

## 0.1.0

  * Initial release.


lib/stream/s3UsageStream.js
/**
 * @fileOverview S3UsageStream class definition.
 */

// Core.
var Transform = require('stream').Transform;
var util = require('util');

// NPM.
var _ = require('lodash');

// Local.
var constants = require('../constants');

//---------------------------------------------------------------------------
// Class constructor.
//---------------------------------------------------------------------------

/**
 * @class An object stream for running totals of size and count of S3 objects.
 *
 * Pipe in S3 object definitions from the S3ListObjectStream, pipe out running
 * total objects with counts and size grouped by bucket and key prefix:
 *
 * [
 *   {
 *     path: 'bucket/folder1',
 *     storageClass: {
 *       STANDARD: {
 *         // Number of objects counted so far.
 *         count: 55,
 *         // Total size of all objects counted so far in bytes.
 *         size: 1232983
 *       },
 *       STANDARD_IA: {
 *         count: 0,
 *         size: 0
 *       },
 *       REDUCED_REDUNDANCY: {
 *         count: 2,
 *         size: 5638
 *       },
 *       GLACIER: {
 *         count: 0,
 *         size: 0
 *       }
 *     }
 *   }
 *   ,
 * ]
 *
 * @param {Object} options Standard stream options, plus those noted.
 * @param {Object} [options.delimiter] How to split keys into folders. Defaults
 *   to '/',
 * @param {Object} [options.depth] Depth of folders to group count and size.
 *   Defaults to 0, or no folders, just buckets.
 * @param {Number} [options.outputFactor] At 1, send an update for every object.
 *   At 10, only once every 10 objects. Defaults to 100. Updating is expensive
 *   for very large buckets or depths.
 */
function S3UsageStream (options) {
  options = options || {};
  // Important; make this an object stream.
  options.objectMode = true;

  this.delimiter = options.delimiter || '/';
  this.depth = options.depth || 0;
  this.outputFactor = options.outputFactor || 100;

  // The running totals.
  this.totals = {};
  this.sortedTotals = [];
  this.count = 0;

  S3UsageStream.super_.call(this, options);
}

util.inherits(S3UsageStream, Transform);

//---------------------------------------------------------------------------
// Methods
//---------------------------------------------------------------------------

/**
 * Update the running totals and return a copy.
 *
 * @param {Object} s3Object An S3 object definition.
 * @param {Function} callback Of the form function (error).
 */
S3UsageStream.prototype.updateTotals = function (s3Object, callback) {
  var self = this;
  var paths = [s3Object.Bucket];
  var pathSegments;
  var index;
  var limit;

  // Future proofing against new storage classes.
  if (!_.includes(constants.storageClass, s3Object.StorageClass)) {
    return callback (new Error(util.format(
      'Unknown storage class %s for S3 object %s.',
      s3Object.StorageClass,
      s3Object.Key
    )));
  }

  // Assemble paths. For something like s3://bucket/a/b/c, depending on depth we
  // want 'bucket', 'bucket/a', 'bucket/a/b', and so on.
  if (this.depth > 0) {
    pathSegments = s3Object.Key.split(this.delimiter);
    // Drop the last segment, which is the file name.
    pathSegments.pop();
    limit = Math.min(this.depth, pathSegments.length);

    for (index = 0; index < limit; index++) {
      paths.push(
        [s3Object.Bucket].concat(pathSegments.slice(0, index + 1)).join(this.delimiter)
      );
    }
  }

  // Run through each of the paths, add the totals.
  paths.forEach(function (path) {
    // If we don't have a record for this path yet, then create one.
    if (!self.totals[path]) {
      var obj = {
        path: path,
        storageClass: {}
      };

      _.each(constants.storageClass, function (storageClass) {
        obj.storageClass[storageClass] = {
          count: 0,
          size: 0
        }
      });
      self.totals[path] = obj;


      var index = _.sortedIndexBy(self.sortedTotals, self.totals[path], function (item) {
        return item.path;
      });

      self.sortedTotals.splice(index, 0, self.totals[path]);
    }

    self.totals[path].storageClass[s3Object.StorageClass].count++;
    self.totals[path].storageClass[s3Object.StorageClass].size += s3Object.Size;
  });

  // Sending on a cloned object is an expensive operation for very large buckets
  // or large depths, meaning a large number of items in the array. Increment
  // the count, and only send on an object every outputFactor times.
  this.count++;

  if (this.count % this.outputFactor === 0) {
    this.push(_.cloneDeep(this.sortedTotals));
  }

  callback();
};

/**
 * Override end() to ensure that at least one output is delivered despite the
 * outputFactor setting. This will mean that there will be a duplicate sent for
 * some circumstances, but that isn't important in this usage.
 */
S3UsageStream.prototype.end = function () {
  this.push(_.cloneDeep(this.sortedTotals));
  S3UsageStream.super_.prototype.end.apply(this, arguments);
};

/**
 * Implementation of the necessary transform method.
 *
 * @param {Object} data An S3 object definition.
 * @param {String} encoding Irrelevant since this is an object stream.
 * @param {Function} callback Invoked after this listing is processed.
 */
S3UsageStream.prototype._transform = function (data, encoding, callback) {
  if (
    !data ||
    typeof data.Bucket !== 'string' ||
    typeof data.Key !== 'string' ||
    typeof data.Size !== 'number'
  ) {
    return callback(new Error('Invalid S3 object definition provided: ' + JSON.stringify(data)));
  }

  this.updateTotals(data, callback);
};

//---------------------------------------------------------------------------
// Export class constructor.
//---------------------------------------------------------------------------

module.exports = S3UsageStream;

